app/assets/javascripts/common/admins.js

```diff
@@ -1,97 +1,117 @@
+//
+// elements
+//
 const ACCESS_LIST_INPUT_SELECTOR = "input.access-input"
-const facilityAccessDiv = () => document.getElementById("facility-access")
-const selectAllFacilitiesDiv = () => document.getElementById("select-all-facilities")
-const facilityAccessItemsAccessRatio = () => document.getElementsByClassName("access-ratio")
-const facilityAccessItemsPadding = () => document.getElementsByClassName("access-item__padding")
-const facilityAccessPowerUser = () => document.getElementById("facility-access-power-user")
+const $facilityAccessDiv = () => document.getElementById("facility-access")
+const $selectAllFacilitiesContainer = () => document.getElementById("select-all-facilities")
+const $selectAllFacilitiesInput = () => document.getElementById("select-all-facilities-input")
+const $facilityAccessItemsAccessRatio = () => document.getElementsByClassName("access-ratio")
+const $facilityAccessItemsPadding = () => document.getElementsByClassName("access-item__padding")
+const $facilityAccessPowerUser = () => document.getElementById("facility-access-power-user")
+const $accessLevel = () => document.getElementById("access_level")
+const $page = () => $facilityAccessDiv() && $facilityAccessDiv().attributes.getNamedItem('data-page').value
+
 //
-// loads at page refresh
+// load things upfront
 //
-window.addEventListener("DOMContentLoaded", inviteAdmin);
-window.addEventListener("DOMContentLoaded", editAdmin);
+window.addEventListener("DOMContentLoaded", onPageLoad);
 
-function inviteAdmin() {
+function onPageLoad() {
   accessLevelListener()
-  selectAccessLevels()
-  selectAllListener()
+  accessLevelSelector()
+}
+
+//
+// wait for the render_async hook before loading the access tree related JS
+//
+document.addEventListener('render_async_load', function (_event) {
+  selectAllButtonListener()
   checkboxItemListener()
   resourceRowCollapseListener()
-}
+  editAdmin()
+});
 
 function editAdmin() {
+  if (!isCurrentPage("edit")) return
+
   // list of all checkboxes under facilityAccessDiv()
-  const checkboxes = nodeListToArray(ACCESS_LIST_INPUT_SELECTOR, facilityAccessDiv())
-  const checkedCheckboxes = checkboxes.filter(check => check.checked)
+  const checkboxes = nodeListToArray(ACCESS_LIST_INPUT_SELECTOR, $facilityAccessDiv())
+
+  // go through all the checkboxes that are pre-checked and update their parents accordingly
+  for (const checkbox of checkboxes) {
+    if (!checkbox.checked) continue
 
-  for (const checkbox of checkedCheckboxes) {
-    updateParentCheckedState(checkbox, ACCESS_LIST_INPUT_SELECTOR)
+    // a large tree can take a lot of time to load on the DOM,
+    // so we queue up our updates by requesting frames so as to not cause overwhelming repaints
+    requestAnimationFrame(function () {
```

Comment:
2/2:

This is the second useful change, when we're in the edit screen, and we pre-check the appropriate check-boxes, we still need to update the parents for indeterminate state ([can't provide indeterminate state from HTML](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox#Indeterminate_state_checkboxes)).

This needs to loop through all the checked checkboxes and apply `updateParentCheckedState` on each, which can be quite expensive and I've seen slow rendering times because of it.

Here we queue up these updates so that it doesn't block the UI thread and make the page slow to render. It works quite well.